controlplane/kubeadm/internal/cluster.go

```diff
@@ -0,0 +1,383 @@
+	selector := map[string]string{
+		clusterv1.ClusterLabelName: cluster.Name,
+	}
+	allMachines := &clusterv1.MachineList{}
+	if err := m.Client.List(ctx, allMachines, client.InNamespace(cluster.Namespace), client.MatchingLabels(selector)); err != nil {
+		return nil, errors.Wrap(err, "failed to list machines")
+	}
+	if len(filters) == 0 {
+		return allMachines.Items, nil
+	}
+	filteredMachines := []clusterv1.Machine{}
+	for _, machine := range allMachines.Items {
+		add := true
+		for _, filter := range filters {
+			if !filter(machine) {
+				add = false
+				break
+			}
+		}
+		if add {
+			filteredMachines = append(filteredMachines, machine)
+		}
+	}
+	return filteredMachines, nil
+}
+
+// getCluster builds a cluster object.
+// The cluster is also populated with secrets stored on the management cluster that is required for
+// secure internal pod connections.
+func (m *ManagementCluster) getCluster(ctx context.Context, clusterKey types.NamespacedName) (*cluster, error) {
+	// This adapter is for interop with the `remote` package.
+	adapterCluster := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: clusterKey.Namespace,
+			Name:      clusterKey.Name,
+		},
+	}
+
+	// TODO(chuckha): Unroll remote.NewClusterClient if we are unhappy with getting a restConfig twice.
+	restConfig, err := remote.RESTConfig(ctx, m.Client, adapterCluster)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := remote.NewClusterClient(ctx, m.Client, adapterCluster, scheme.Scheme)
+	if err != nil {
+		return nil, err
+	}
+	etcdCACert, etcdCAKey, err := m.GetEtcdCerts(ctx, clusterKey)
+	if err != nil {
+		return nil, err
+	}
+	return &cluster{
+		client:     c,
+		restConfig: restConfig,
+		etcdCACert: etcdCACert,
+		etcdCAkey:  etcdCAKey,
+	}, nil
+}
+
+// GetEtcdCerts returns the EtcdCA Cert and Key for a given cluster.
+func (m *ManagementCluster) GetEtcdCerts(ctx context.Context, cluster types.NamespacedName) ([]byte, []byte, error) {
+	etcdCASecret := &corev1.Secret{}
+	etcdCAObjectKey := types.NamespacedName{
+		Namespace: cluster.Namespace,
+		Name:      fmt.Sprintf("%s-etcd", cluster.Name),
+	}
+	if err := m.Client.Get(ctx, etcdCAObjectKey, etcdCASecret); err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to get secret; etcd CA bundle %s/%s", etcdCAObjectKey.Namespace, etcdCAObjectKey.Name)
+	}
+	return etcdCASecret.Data["tls.crt"], etcdCASecret.Data["tls.key"], nil
+}
+
+// TargetClusterEtcdIsHealthy runs a series of checks over a target cluster's etcd cluster.
+// In addition, it verifies that there are the same number of etcd members as control plane Machines.
+func (m *ManagementCluster) TargetClusterEtcdIsHealthy(ctx context.Context, clusterKey types.NamespacedName, controlPlaneName string) error {
+	cluster, err := m.getCluster(ctx, clusterKey)
+	if err != nil {
+		return err
+	}
+	resp, err := cluster.etcdIsHealthy(ctx)
+	if err != nil {
+		return err
+	}
+	errorList := []error{}
```

Comment:
Interesting, is this a style preference? 

It doesn't appear to be convention because slightly farther down in the cluster_controller we're doing this: https://github.com/kubernetes-sigs/cluster-api/blob/5d3d0d1cbfda7a6aa101834509e1ac402ecd0ce1/controllers/cluster_controller.go#L160-L174

Reply:
I hadn't seen it before working on the webhook conversions to CAPA, so maybe a new(ish) convention, and we should mop up elsewhere.

Reply:
We can probably take care of it later

Reply:
I'm not sure we should adopt that convention outside of `defer`, otherwise we risk overwriting or otherwise mishandling the aggregated err var.